# Exercises

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.formula.api as smf
import statsmodels.stats.power as smp
import statsmodels.stats.proportion as smprop
import matplotlib_inline.backend_inline
matplotlib_inline.backend_inline.set_matplotlib_formats('svg')

# Exercises

## 3.5 Transport times

A company, MM, selling items online wants to compare the transport times for two transport firms for delivery of the goods. To compare the two companies recordings of delivery times on a specific route were made, with a sample size of $n = 9$ for each firm. The following data were found:

Firm A: $\bar{y}_A = 1.93$ d and $s_A = 0.45$ d,

Firm B: $\bar{y}_B = 1.49$ d and $s_B = 0.58$ d.

note that d is the SI unit for days. It is assumed that data can be regarded as stemming from normal distributions.

### a)
We want to test the following hypothesis

$$H_0 : \mu_A = \mu_B$$
$$H_1 : \mu_A \neq \mu_B$$

What is the p-value, interpretation and conclusion for this test (at $\alpha = 5\%$ level)?

**Solution:**

In [2]:
y_A, s_A, n_A = 1.93, 0.45, 9
y_B, s_B, n_B = 1.49, 0.58, 9

t_obs, p_value = stats.ttest_ind_from_stats(
    mean1=y_A, std1=s_A, nobs1=n_A,
    mean2=y_B, std2=s_B, nobs2=n_B,
    equal_var=False  # Welch's t-test
)

print(f"t_obs:   {t_obs:.4f}")
print(f"p-value: {p_value:.4f}")

t_obs:   1.7981
p-value: 0.0922


### b)
Find the 95% confidence interval for the mean difference $\mu_A - \mu_B$.

**Solution:**

In [3]:
y_A, s_A, n_A = 1.93, 0.45, 9
y_B, s_B, n_B = 1.49, 0.58, 9

se = np.sqrt(s_A**2/n_A + s_B**2/n_B)
nu = (s_A**2/n_A + s_B**2/n_B)**2 / ((s_A**2/n_A)**2/(n_A-1) + (s_B**2/n_B)**2/(n_B-1))

t_crit = stats.t.ppf(0.975, df=nu)
diff = y_A - y_B
ci_lower = diff - t_crit * se
ci_upper = diff + t_crit * se

print(f"nu = {nu:.4f}")
print(f"95% CI for mu_A - mu_B: [{ci_lower:.4f}, {ci_upper:.4f}]")

nu = 15.0696
95% CI for mu_A - mu_B: [-0.0814, 0.9614]


### c)
What is the power of a study with $n = 9$ observations in each of the two samples of detecting a potential mean difference of 0.4 between the firms (assume that $\sigma = 0.5$ and that we use $\alpha = 0.05$)?

**Solution:**

In [4]:
delta = 0.4
sigma = 0.5
alpha = 0.05
n = 9

effect_size = delta / sigma
power = smp.TTestIndPower().solve_power(effect_size=effect_size, nobs1=n, alpha=alpha, ratio=1)
print(f"Power: {power:.4f}")

Power: 0.3579


### d)
What effect size (mean difference) could be detected with $n = 9$ observations in each of the two samples with a power of 0.8 (assume that $\sigma = 0.5$ and that we use $\alpha = 0.05$)?

**Solution:**

In [5]:
effect_size = smp.TTestIndPower().solve_power(nobs1=n, alpha=alpha, power=power, ratio=1)
delta = effect_size * sigma

print(f"Effect size: {effect_size:.4f}")
print(f"Detectable delta: {delta:.4f}")

Effect size: 0.8000
Detectable delta: 0.4000


### e)
How large a sample size (from each firm) would be needed in a new investigation, if we want to detect a potential mean difference of 0.4 between the firms with probability 0.90, that is with power=0.90 (assume that $\sigma = 0.5$ and that we use $\alpha = 0.05$)?

**Solution:**

In [6]:
delta = 0.4
sigma = 0.5
alpha = 0.05
power = 0.90

effect_size = delta / sigma
n = smp.TTestIndPower().solve_power(effect_size=effect_size, alpha=alpha, power=power, ratio=1)
print(f"Required n per group: {n:.4f}")

Required n per group: 33.8255


## 3.6 Cholesterol

In a clinical trial of a cholesterol-lowering agent, 15 patients' cholesterol (in mmol/L) has been measured before treatment and 3 weeks after starting treatment. Data are listed in the following table:

| Patient | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Before | 9.1 | 8.0 | 7.7 | 10.0 | 9.6 | 7.9 | 9.0 | 7.1 | 8.3 | 9.6 | 8.2 | 9.2 | 7.3 | 8.5 | 9.5 |
| After | 8.2 | 6.4 | 6.6 | 8.5 | 8.0 | 5.8 | 7.8 | 7.2 | 6.7 | 9.8 | 7.1 | 7.7 | 6.0 | 6.6 | 8.4 |

The following is run in Python:
```python
x1 = np.array([9.1, 8.0, 7.7, 10.0, 9.6, 7.9, 9.0, 7.1,
               8.3, 9.6, 8.2, 9.2, 7.3, 8.5, 9.5])
x2 = np.array([8.2, 6.4, 6.6, 8.5, 8.0, 5.8, 7.8, 7.2,
               6.7, 9.8, 7.1, 7.7, 6.0, 6.6, 8.4])

# Independent
t_obs, p_val = stats.ttest_ind(x1, x2, equal_var=False)
print(t_obs, p_val)
```
```
3.3206371615181602 0.0025776851099507973
```
```python
# Paired
t_obs, p_val = stats.ttest_rel(x1, x2)
print(t_obs, p_val)
```
```
7.340653674886342 3.6722518645935146e-06
```

### a)
Can there, based on these data, be demonstrated a significant decrease in cholesterol levels with $\alpha = 0.001$?

**Solution:**

In [7]:
x1 = np.array([9.1, 8.0, 7.7, 10.0, 9.6, 7.9, 9.0, 7.1,
               8.3, 9.6, 8.2, 9.2, 7.3, 8.5, 9.5])
x2 = np.array([8.2, 6.4, 6.6, 8.5, 8.0, 5.8, 7.8, 7.2,
               6.7, 9.8, 7.1, 7.7, 6.0, 6.6, 8.4])

alpha = 0.001
t_obs, p_val = stats.ttest_rel(x1, x2)
print(f"t_obs:   {t_obs:.4f}")
print(f"p-value: {p_val:.6e}")
print(f"Reject H0: {p_val < alpha}")

t_obs:   7.3407
p-value: 3.672252e-06
Reject H0: True


## 3.8 Foil production

In the production of a certain foil (film), the foil is controlled by measuring the thickness of the foil in a number of points distributed over the width of the foil. The production is considered stable if the mean of the difference between the maximum and minimum measurements does not exceed 0.35 mm. At a given day, the following random samples are observed for 10 foils:

| Foil | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
|---|---|---|---|---|---|---|---|---|---|---|
| Max. in mm ($y_{max}$) | 2.62 | 2.71 | 2.18 | 2.25 | 2.72 | 2.34 | 2.63 | 1.86 | 2.84 | 2.93 |
| Min. in mm ($y_{min}$) | 2.14 | 2.39 | 1.86 | 1.92 | 2.33 | 2.00 | 2.25 | 1.50 | 2.27 | 2.37 |
| Max-Min (D) | 0.48 | 0.32 | 0.32 | 0.33 | 0.39 | 0.34 | 0.38 | 0.36 | 0.57 | 0.56 |

The following statistics may potentially be used

$$\bar{y}_{max} = 2.508, \quad \bar{y}_{min} = 2.103, \quad s_{y_{max}} = 0.3373, \quad s_{y_{min}} = 0.2834, \quad s_D = 0.09664.$$

### a)
What is a 95% confidence interval for the mean difference?

**Solution:**

In [10]:
y_max = np.array([2.62, 2.71, 2.18, 2.25, 2.72, 2.34, 2.63, 1.86, 2.84, 2.93])
y_min = np.array([2.14, 2.39, 1.86, 1.92, 2.33, 2.00, 2.25, 1.50, 2.27, 2.37])
D = np.array([0.48, 0.32, 0.32, 0.33, 0.39, 0.34, 0.38, 0.36, 0.57, 0.56])
y_max_bar = 2.508
y_min_bar = 2.103
s_y_max = 0.3373
s_y_min = 0.2834
s_D = 0.09664
n = len(y_max)

alpha = 0.05
d_mean = D.mean()

ci = stats.t.interval(1 - alpha, df=n-1, loc=d_mean, scale=s_D/np.sqrt(n))
print(f"d_mean: {d_mean:.4f}")
print(f"95% CI for mean difference: [{ci[0]:.4f}, {ci[1]:.4f}]")

d_mean: 0.4050
95% CI for mean difference: [0.3359, 0.4741]


### b)
How much evidence is there that the mean difference is different from 0.35? State the null hypothesis, t-statistic and p-value for this question.

**Solution:**

In [11]:
t_obs = d_mean / (s_D / np.sqrt(n))
p_value = 2 * (1 - stats.t.cdf(abs(t_obs), df=n-1))

print(f"t_obs:   {t_obs:.4f}")
print(f"p-value: {p_value:.4f}")
print(f"Reject H0: {p_value < alpha}")

t_obs:   13.2525
p-value: 0.0000
Reject H0: True


## 3.10 Concrete items (sample size)

This is a continuation of Exercise 3.1, so the same setting and data is used (read the initial text of it).

### a)
A study is planned of a new supplier. It is expected that the standard deviation will be approximately 3, that is, $\sigma = 3$ mm. We want a 90% confidence interval for the mean value in this new study to have a width of 2 mm. How many items should be sampled to achieve this?

**Solution:**

In [17]:
mu_0 = 0.35
alpha = 0.1

t_obs = (d_mean - mu_0) / (s_D / np.sqrt(n))
p_value = 2 * (1 - stats.t.cdf(abs(t_obs), df=n-1))

print(f"H0: mu_D = 0.35")
print(f"H1: mu_D != 0.35")
print(f"t_obs:   {t_obs:.4f}")
print(f"p-value: {p_value:.6f}")
print(f"Reject H0: {p_value < alpha}")

H0: mu_D = 0.35
H1: mu_D != 0.35
t_obs:   1.7074
p-value: 0.126138
Reject H0: False


### b)
Answer the sample size question above but requiring the 99% confidence interval to have the (same) width of 2 mm.

**Solution:**

In [18]:
alpha = 0.01
mu_0 = 0.35

t_obs = (d_mean - mu_0) / (s_D / np.sqrt(n))
p_value = 2 * (1 - stats.t.cdf(abs(t_obs), df=n-1))

print(f"H0: mu_D = 0.35")
print(f"H1: mu_D != 0.35")
print(f"t_obs:   {t_obs:.4f}")
print(f"p-value: {p_value:.6f}")
print(f"Reject H0: {p_value < alpha}")

H0: mu_D = 0.35
H1: mu_D != 0.35
t_obs:   1.7074
p-value: 0.126138
Reject H0: False


### c)
(Warning: This is a difficult question about a challenging abstraction - do not worry, if you do not make this one) For the two sample sizes found in the two previous questions find the probability that the corresponding confidence interval in the future study will actually be more than 10% wider than planned for (still assuming and using that the population variance is $\sigma^2 = 9$).

**Solution:**

### d)
Now a new experiment is to be planned. In the first part above, given some wanted margin of error (ME) a sample size of $n = 25$ was found. What are each of the probabilities that an experiment with $n = 25$ will detect effects corresponding to ("end up significant for") $\mu_1 = 3001, 3002, 3003$ respectively? Assume that we use the typical $\alpha = 0.05$ level and that $\sigma = 3$?

**Solution:**

### e)
One of the sample size computation above led to $n = 60$ (it is not so important how/why). Answer the same question as above using $n = 60$.

**Solution:**

### f)
What sample size would be needed to achieve a power of 0.80 for an effect of size 0.5?

**Solution:**

### g)
Assume that you only have the finances to do an experiment with $n = 50$. How large a difference would you be able to detect with probability 0.8 (i.e. Power = 0.80)?

**Solution:**

## E9.10 Proof of Corollary 9.29

In this exercise we will prove Corollary 9.29 by a series of sub questions.

### a)
Show that if $Y \sim N(X\beta, \sigma^2 I)$ then

$$Y^T(I - H_1)Y \sim \chi^2(n-2). \tag{9-2}$$

Independently of the value of $\beta$

**Solution:**

### b)
Show that if $Y \sim N(\mathbf{1}\mu, \sigma^2 I)$ then

$$Y^T(H_1 - H_0)Y \sim \chi^2(1). \tag{9-3}$$

independently of the value of $\mu$, you may use the formulation in (9-137) to calculate $H_1$, or simply use the fact that $\mathbf{1}H_1 = \mathbf{1}^T$ (see Exercise 9.11).

**Solution:**

### c)
Show that if $Y \sim N(0, \sigma^2 I)$ then

$$Y^T H_0 Y \sim \chi^2(1). \tag{9-4}$$

**Solution:**

## E9.11 t-test Orthogonal projections

### a)
Show that the projection matrices in rhs of (9-141) are orthogonal i.e. $H_0(H_1 - H_0) = 0$, $H_0(I - H_1) = 0$ and $(H_1 - H_0)(I - H_1) = 0$. Hint: you may start by showing that $X_0^T H_1 = X_0^T$. You may use the parametrization (9-137).

**Solution:**

### b)
Use the result to show that

$$\text{Cov}[H_0 Y, (H_1 - H_0)Y] = 0$$
$$\text{Cov}[H_0 Y, (I - H_1)Y] = 0$$
$$\text{Cov}[(H_1 - H_0)Y, (I - H_1)Y] = 0 \tag{9-5}$$

and hence that the projected vectors are independent. Also what is the interpretation in terms of fitted values?

**Solution:**

## E9.13 t-test Central estimators under Null-hypothesis

Consider the projection matrices for the two sample t-test (equation (9-141)), consider two groups $Y_{1,i} \sim N(\mu_1, \sigma^2)$ and iid., $i = \{1, 2, ..., n_1\}$ and $Y_{2,j} \sim N(\mu_2, \sigma^2)$ and iid., $j = \{1, 2, ..., n_2\}$. Define $Y = [Y_1^T, Y_2^T]^T = [Y_{1,1}, ..., Y_{1,n_1}, Y_{2,1}, ..., Y_{2,n_2}]^T$ and

### a)
Show that

$$Y^T(H_1 - H_0)Y = \frac{n_1 n_2}{n_1 + n_2}(\bar{Y}_1 - \bar{Y}_2)^2 \tag{9-6}$$

**Solution:**

### b)
Show that $E[Y^T(H_1 - H_0)Y] = \frac{n_1 n_2}{n_1+n_2}(\mu_1 - \mu_2)^2 + \sigma^2$

Under the assumption $\mu_1 = \mu_2 = \mu$ conclude that $Y^T(H_1 - H_0)Y$ is a central estimator for $\sigma^2$, find the variance of this estimator, and compare with the estimator (9-147).

**Solution:**